src/iceberg/type.h

```diff
@@ -0,0 +1,149 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+/// \file iceberg/type.h
+/// Data types for Iceberg.
+
+#include <memory>
+#include <span>
+#include <string>
+#include <vector>
+
+#include "iceberg/iceberg_export.h"
+#include "iceberg/type_fwd.h"
+
+namespace iceberg {
+
+/// \brief A type combined with a name.
+class ICEBERG_EXPORT Field {
```

Comment:
I think similarly to Schema, I'd put Field into a separate file. Now this is in type.h but Field isn't really a type.

Reply:
There's a bit of a circular dependency between Field (which wants Type) and StructType (which wants Field). Unless (again) you see the alternate design in the first two comments of this PR which puts child Fields always on the Field (this means that Field has a potentially redundant...field, but avoids nesting of types - I personally slightly prefer this but it's conceptually dependent on whether you model the type itself as nested or the field as the nested object)

Reply:
Sure, will take a look a bit later.
Generally asking, can't we solve the circular dependency with forward declarations and using pointers/references?

Reply:
We can if we continue to type-erase Type, but not if we want it to not be behind a smart pointer.

Reply:
Putting them behind smart pointers makes sense. See my other comment. And the this could go into a separate file.
Meanwhile I'm sketching some code for the PartitionSpec and PartitionField and I'm wondering if this one here is intuitive to be called Field, or could be better called SchemaField (considering there is going to be a PartitionField too).

Reply:
I'll rename

Reply:
Do you suppose `PartitionField` to inherit `SchemaField` or it is totally a separate thing? @gaborkaszab